### PR TITLE
[narwhal] shutdown network when reconfigure

### PR DESCRIPTION
**Pull request by akichidis**
This PR is taking care of a few things:

* explicitly shutting down primary/worker's network when reconfiguration (Shutdown) happens
* convert the signal propagation from primary -> worker to a reliable broadcast (although that part will effectively be removed in later iterations)
* refactor the `restart` test in order to ensure that a node left behind in epoch changes can catch up with others and can be reconfigured to the latest epoch that other nodes did.

The `restart` test as is won't work correctly , as the `restarter` handler will need to be refactor to allow us pass a new `registry` every time we reconfigure. Once this work is done https://mysten.atlassian.net/browse/SUI-1028 then the `restarter` can be refactored to make it work

### CODE REVIEW

File: narwhal/primary/src/state_handler.rs [hunk 1]
```diff
@@ -129,7 +133,7 @@ impl StateHandler {
             .map(|info| info.name)
             .collect();
 
-        self.network.unreliable_broadcast(our_workers, &message);
+        self.network.broadcast(our_workers, &message)
```
Comment (huitseeker) [inline narwhal/primary/src/state_handler.rs:136]: Should this change to a retrying message be done here or offered in a new `reliably_notify_our_workers` method?

File: narwhal/primary/src/state_handler.rs [hunk 2]
```diff
@@ -166,10 +170,29 @@ impl StateHandler {
...
                         .expect("Reconfigure channel dropped");
 
+                    warn!("Waiting to broadcast reconfigure message to workers");
+
+                    // wait for all the workers to eventually receive the message
+                    // TODO: this request will be removed https://mysten.atlassian.net/browse/SUI-984
+                    let join_all = futures::future::try_join_all(notify_handlers);
+                    join_all.await.expect("Error while sending reconfiguration message to the workers");
+
+                    warn!("Successfully broadcasted reconfigure message to workers");
```
Comment (huitseeker) [inline narwhal/primary/src/state_handler.rs:180]: Nit
```suggestion
                    warn!("Successfully broadcast reconfigure message to workers");
```

File: narwhal/node/tests/reconfigure.rs [hunk 3]
```diff
@@ -97,37 +96,42 @@ impl ExecutionState for SimpleExecutionState {
...
+
+    async fn send_new_epoch_reconfigure(&self, epoch: Epoch) {
+        {
+            let mut guard = self.committee.lock().unwrap();
+            guard.epoch = epoch;
+        }
+
+        let worker_keypairs = self.worker_keypairs.iter().map(|kp| kp.copy());
+        let worker_ids = 0..self.worker_keypairs.len() as u32;
+        let worker_ids_and_keypairs = worker_ids.zip(worker_keypairs).collect();
```
Comment (huitseeker) [inline narwhal/node/tests/reconfigure.rs:120]:
```suggestion
        let worker_ids_and_keypairs = worker_keypair.iter().cloned().enumerate().collect();
```

File: narwhal/node/tests/reconfigure.rs [hunk 4]
```diff
@@ -253,22 +268,62 @@ async fn restart() {
...
+
+            // Repeatedly send transactions.
+            let mut interval = interval(Duration::from_secs(3));
+            interval.set_missed_tick_behavior(MissedTickBehavior::Skip);
+            tokio::pin!(interval);
+
+            loop {
+                tokio::select! {
+                    result = rx.recv() => {
+                        // channel broke - we need to exit the loop
```
Comment (huitseeker) [inline narwhal/node/tests/reconfigure.rs:286]: Nit: The wording "broke" is perhaps not ideal? 
> This method returns None if the channel has been closed and there are no remaining messages in the channel’s buffer
https://docs.rs/tokio/latest/tokio/sync/mpsc/struct.Receiver.html

File: narwhal/primary/src/state_handler.rs (hunk 1 above)
Comment (akichidis) [reply narwhal/primary/src/state_handler.rs:136]: To be honest I tried to make the minimum amount of work here as we are going to remove this call between primary -> worker anyways . Is the `reliably_notify_our_workers` method you suggest part of the state handler or make it part of the network?

File: narwhal/node/tests/reconfigure.rs (hunk 3 above)
Comment (akichidis) [reply narwhal/node/tests/reconfigure.rs:120]: Ah great reduction - I didn't pay too much attention to this

File: narwhal/node/tests/reconfigure.rs (hunk 4 above)
Comment (akichidis) [reply narwhal/node/tests/reconfigure.rs:286]: good point, I'll refactor the comment

File: narwhal/primary/src/state_handler.rs [hunk 5]
```diff
@@ -145,7 +149,7 @@ impl StateHandler {
 
                 Some(message) = self.rx_state_handler.recv() => {
                     // Notify our workers
-                    self.notify_our_workers(message.to_owned());
+                    let notify_handlers = self.notify_our_workers(message.to_owned());
```
Comment (bmwill) [inline narwhal/primary/src/state_handler.rs:152]: Do we know what the semantics around this are supposed to be? is a successful response mean the worker has properly shutdown? How does that work if the network we're sending this message on is intended to be shutdown as well?

Comment (akichidis) [reply narwhal/primary/src/state_handler.rs:152]: To be honest this is a quite fragile as logic to begin with and that's why we want to remove the call anyways (FYI https://mysten.atlassian.net/browse/SUI-984) and instead follow an approach where we'll manage the primary/worker shutdown independently as part of the upcoming iterations. So I don't feel confidence on the approach in either case (reliable vs unreliable send).

Now , for the current semantics I expect for the worker to manage to respond before shutting down - but I want to clarify that we don't have strong guarantees here. Maybe we could do some very hackish stuff in the worker to delay the propagation of the reconfiguration signal by some time, so we can get more confidence that the endpoint will manage to return .

Comment (bmwill) [reply narwhal/primary/src/state_handler.rs:152]: Yeah we can't communicate this signal over an interface that is going to be shutdown, but it sounds like you already have plans to address that

File: narwhal/node/src/restarter.rs [hunk 6]
```diff
@@ -85,6 +85,10 @@ impl NodeRestarter {
             handles.extend(primary_handles);
             handles.extend(worker_handles);
 
+            // give some time to the node to bootstrap before we are ready to receive
+            // another reconfiguration message
+            tokio::time::sleep(tokio::time::Duration::from_secs(2)).await;
```
Comment (asonnino) [inline narwhal/node/src/restarter.rs:90]: why is this needed?

File: narwhal/node/tests/reconfigure.rs [hunk 7]
```diff
@@ -97,37 +96,45 @@ impl ExecutionState for SimpleExecutionState {
...
+        self.tx_reconfigure
+            .send((
+                self.keypair.copy(),
+                self.network_keypair.copy(),
+                new_committee,
+                worker_ids_and_keypairs,
+                self.worker_cache.clone(),
+            ))
+            .await
+            .unwrap();
```
Comment (asonnino) [inline narwhal/node/tests/reconfigure.rs:136]: are we sure this will never panic upon shutdown ?

File: narwhal/primary/tests/epoch_change.rs [hunk 8]
```diff
@@ -122,13 +147,17 @@ async fn test_simple_epoch_change() {
                 .send();
             futs.push(fut);
         }
+
+        info!("Sent to all the reconfigure message {:?}", message);
```
Comment (asonnino) [inline narwhal/primary/tests/epoch_change.rs:151]:
```suggestion
        info!("Sent to all the reconfigure message {message:?}");
```

**Review by asonnino — APPROVED** [review]

File: narwhal/node/src/restarter.rs (hunk 6 above)
Comment (akichidis) [reply narwhal/node/src/restarter.rs:90]: When we bootstrap the node we don't have strong signal whether the node has been fully bootstrapped or there are still components that try to start. That give us some slack in. case we immediately try to reconfigure epochs back to back - mostly an issue observed in tests. In the upcoming iterations will try to introduce readiness signals when components bootstrap
